src/Interpreters/ColumnAliasesVisitor.h

```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include <Interpreters/Aliases.h>
+#include <Interpreters/InDepthNodeVisitor.h>
+#include <Storages/ColumnsDescription.h>
+
+namespace DB
+{
+
+class IAST;
+using ASTPtr = std::shared_ptr<IAST>;
+class IDataType;
+using DataTypePtr = std::shared_ptr<const IDataType>;
+
+/// Visits AST node to rewrite alias columns in filter query
+/// Currently works only in `KeyCondition` of select query and `required_columns` in `InterpreterSelectQuery.cpp`
```

Comment:
Maybe worth add some examples, i.e.:
- alias on alias (required_columns)
- partition pruning (KeyCondition)

Also it is unclear will it help for things like optimize_aggregation_in_order/optimize_read_in_order, looks like the answer is no? (since it replaces only in WHERE/PREWHERE?)

Reply:
you are right.